mne/rank.py

```diff
@@ -344,7 +344,7 @@ def compute_rank(inst, rank=None, scalings=None, info=None, tol='auto',
 
     simple_info = _simplify_info(info)
     picks_list = _picks_by_type(info, meg_combined=True, ref_meg=False,
-                                exclude='bads')
+                                exclude=[])
```

Comment:
In most cases I think we don't want to include bad channels actually, so this seems dangerous to me, assuming it does what I'm assuming it does (and I might be wrong!).

In a practical case, I think for 306-channel neuromag data with no projs for example you should have both on `main` and after this PR:
```
raw = mne.io.read_raw_fif(...)
raw.info['bads'] = []
raw.pick('meg')
assert len(raw.ch_names) == 306
assert compute_rank(raw) == dict(mag=102, grad=204)
cov = compute_raw_covariance(raw)
assert compute_rank(cov) == dict(mag=102, grad=204)
whitener = compute_whitener(cov, pca=True)[0]
assert whitener.shape[0] == 306
```
And then if you mark one channel bad, rank decreases by one:
```
# mark one mag bad, rank decreases by one
raw.info['bads'] = ['MEG 0111']
assert compute_rank(raw) == dict(mag=101, grad=204)
cov = compute_raw_covariance(raw)
assert compute_rank(cov) == dict(mag=101, grad=204)
whitener = compute_whitener(cov, pca=True)[0]
assert whitener.shape[0] == 305
```
Is it still the case after this PR? Either way, please add this sort of thing as an explicit test in `test_rank.py` or (maybe better) `test_cov.py`. 

If the above test is (still) the case and these ranks don't decrease when you pass an explicit `picks=np.arange(306)` for example, then that seems reasonable to me. In other words, I guess this could be an addition to the newly proposed test:
```
# explicit picks restore the old rank
whitener = compute_whitener(cov, pca=True, picks=np.arange(306))[0]
assert whitener.shape[0] == 306
```
Maybe the correct fix is to provide a `picks` argument to `compute_rank` that behaves the same way that it's supposed to in `compute_whitener`, i.e., `'bads'` are ignored by default (`picks_good_data_noref` just like `compute_whitener`) but if you pass picks explicitly then they can be included? That way the above test could also then have after marking one channel bad:
```
assert compute_rank(raw, picks=np.arange(306)) == dict(mag=102, grad=204)
```

Reply:
I agree that this PR should not change the default behavior, i.e. excluding bads. The goal of the PR is to impact only the behavior when explicit picks (bads or not) are provided. From just looking at the code, and the testing I've done until now, the modification should work as intended and should not have impacted the default behavior. However, I didn't have time today to check if the tests were already covering this correctly. I will check tomorrow, and add the additional tests if needed.